include/envoy/stats/timespan.h

```diff
@@ -11,11 +11,25 @@ namespace Envoy {
 namespace Stats {
 
 /**
- * An individual timespan that flushes its measured value (in milliseconds) to a histogram. The
- * initial time is captured on construction. A timespan must be completed via complete() for it to
- * be stored. If the timespan is deleted this will be treated as a cancellation.
+ * An abstraction of timespan which can be completed.
  */
-class Timespan {
+class CompletableTimespan {
+public:
+  virtual ~CompletableTimespan() {}
+
+  /**
+   * Complete the timespan.
+   */
+  virtual void complete() PURE;
+};
+
+/**
+ * An individual timespan that flushes its measured value in time unit (e.g
+ * std::chrono::milliseconds). The initial time is captured on construction. A timespan must be
+ * completed via complete() for it to be stored. If the timespan is deleted this will be treated as
+ * a cancellation.
+ */
+template <class TimeUnit> class Timespan : public CompletableTimespan {
```

Comment:
In general we are going to end up using milliseconds almost everywhere. To reduce the size of this change and make it more convenient to users, what do you think about:
1) Renaming this template class `TimespanWithUnit` (or similar)
2) Having a using declaration that defines `Timespan = TimespanWithUnit<std::chrono::milliseconds>`
3) Then I think the only updated code needs to be in the redis code and here?

Reply:
Thanks @mattklein123 for reviewing this. Makes sense to reduce the number of changes as the default time unit is milliseconds almost everywhere. I initially thought of keeping it in a generalized way where user when instantiating knows what time unit he is using. But for convenience keeping Timespan as milliseconds specialization makes usage easy. I will update the PR soon.

Reply:
@mattklein123 done.